src/main/java/org/duckdb/DuckDBConnection.java

```diff
@@ -110,15 +112,32 @@ protected void finalize() throws Throwable {
         close();
     }
 
-    public synchronized void close() throws SQLException {
-        if (conn_ref != null) {
+    public void close() throws SQLException {
+        if (conn_ref == null) {
+            return;
+        }
+        synchronized (this) {
+            if (conn_ref == null) {
+                return;
+            }
+            // Closing remaining statements is not required by JDBC spec,
+            // but it is a reasonable expectation from clients point of view.
+            List<DuckDBPreparedStatement> stmtList = new ArrayList<>(statements);
+            for (DuckDBPreparedStatement stmt : stmtList) {
+                stmt.close();
+            }
             DuckDBNative.duckdb_jdbc_disconnect(conn_ref);
             conn_ref = null;
         }
     }
 
     public boolean isClosed() throws SQLException {
-        return conn_ref == null;
+        if (conn_ref == null) {
+            return true;
+        }
+        synchronized (this) {
```

Comment:
The use of "synchronized" might cause thread-pinning when used with virtual threads (atleast for java versions from 19 to 23). Would it be better to use a ReentrantLock instead ?

Similar discussion in pgjdbc: https://github.com/pgjdbc/pgjdbc/issues/1951

Reply:
M, thread pinning here is either very short (when `synchronized`` block only flips the flag) or unavoidable, when it goes to a native call. Unlike Postgres, there is no IO done from Java in DuckDB - the pinned thread is used to do the actual work in DB engine code. At the same time, `ReentrantLock` s (with a few `volatile` fields) should not be in any way worse than `synchronized` blocks. So perhaps it makes sense to use them consistently instead of `synchronized`.